target_chains/starknet/contracts/src/wormhole.cairo

```diff
@@ -1,29 +1,370 @@
+            i += 1;
+        };
+        result?;
+        self.current_guardian_set_index.write(set_index);
+        Result::Ok(())
+    }
+
+    fn expire_guardian_set(ref self: ContractState, set_index: u32, now: u64) {
+        let mut set = self.guardian_sets.read(set_index);
+        set.expiration_time = now + 86400;
+        self.guardian_sets.write(set_index, set);
     }
 
     #[abi(embed_v0)]
-    impl HelloImpl of super::IWormhole<ContractState> {
-        fn get_value(self: @ContractState) -> felt252 {
-            self.name.read() + 2
+    impl WormholeImpl of IWormhole<ContractState> {
+        fn submit_new_guardian_set(
+            ref self: ContractState, set_index: u32, guardians: Array<felt252>
+        ) -> Result<(), SubmitNewGuardianSetError> {
+            let execution_info = get_execution_info().unbox();
+            if self.owner.read() != execution_info.caller_address {
+                return Result::Err(SubmitNewGuardianSetError::AccessDenied);
+            }
+            let current_set_index = self.current_guardian_set_index.read();
+            if set_index != current_set_index + 1 {
+                return Result::Err(SubmitNewGuardianSetError::InvalidGuardianSetSequence.into());
+            }
+            expire_guardian_set(
+                ref self, current_set_index, execution_info.block_info.unbox().block_timestamp
+            );
+            store_guardian_set(ref self, set_index, guardians)
         }
 
-        fn set_value(ref self: ContractState, name: felt252) {
-            self.name.write(name - 2);
+        fn parse_and_verify_vm(
+            ref self: ContractState, encoded_vm: ByteArray
+        ) -> Result<VM, ParseAndVerifyVmError> {
+            let (vm, body_hash) = parse_vm(encoded_vm)?;
+            let guardian_set = self.guardian_sets.read(vm.guardian_set_index);
+            if guardian_set.num_guardians == 0 {
+                return Result::Err(ParseAndVerifyVmError::InvalidGuardianSetIndex);
+            }
+            if vm.guardian_set_index != self.current_guardian_set_index.read()
+                && guardian_set.expiration_time < get_block_timestamp() {
+                return Result::Err(ParseAndVerifyVmError::GuardianSetExpired);
+            }
+            if vm.signatures.len() < quorum(guardian_set.num_guardians) {
+                return Result::Err(ParseAndVerifyVmError::NoQuorum);
+            }
+            let mut signatures_clone = vm.signatures.clone();
+            let mut last_index = Option::None;
+
+            let mut result = Result::Ok(());
+            loop {
+                let signature = match signatures_clone.pop_front() {
+                    Option::Some(v) => { v },
+                    Option::None => { break; },
+                };
+
+                match last_index {
+                    Option::Some(last_index) => {
+                        if *(@signature).guardian_index <= last_index {
```

Comment:
What's the reason for the snapshotting and immediate desnapping in this function? Just curious for my own Cairo knowledge.

Reply:
It's not immediate desnapping, it should be read as `*((@signature).guardian_index)`. It forces the compiler to take a snapshot of `signature` when accessing the field. It's a workaround for an issue where the compiler thinks `signature` is moved out, despite `guardian_index` being `Copy`. Not sure if it's a bug or just a limitation of the language.